### PR TITLE
fix(nuxt3): allow customising page keys

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/nuxt/framework/issues/2309

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR removes the default keying of routes to the full path, meaning that parent routes won't rerender when child routes change. It also adds support for setting custom keys for child routes, either a string or a function that receives the current route.

The key can either be set in `definePageMeta` or via a prop for `<NuxtNestedPage />`.

Contrary to [my earlier RFC](https://github.com/nuxt/framework/discussions/2468), I now think it makes sense to set the page key via meta as that way we can implement logic in a key function that will handle both child routes and deeply nested child routes (see following note).

**Note**: nested routes which have children themselves [behaved differently in Nuxt 2](https://github.com/nuxt/nuxt.js/blob/a2495a68736f9af96cd25209a948beb1ca665d4d/packages/vue-app/template/components/nuxt.js#L47-L72). I can restore this behaviour if desired.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt3/src/pages/runtime/nested-page.vue [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
+  <RouterView v-slot="{ Component, route }">
+    <component :is="Component" :key="getKeyFor(childKey || route.meta.key, route)" />
   </RouterView>
 </template>
 
-<script>
+<script lang="ts">
+import type { RouteLocationNormalizedLoaded } from 'vue-router'
+
+const getKeyFor = (key: string | ((route: RouteLocationNormalizedLoaded) => string), route?: RouteLocationNormalizedLoaded) => {
```
Comment (pi0) [inline packages/nuxt3/src/pages/runtime/nested-page.vue:10]: Code style: Can't we merge logic of L3 + `getKeyFor` as a computed prop in the setup?

File: packages/nuxt3/src/pages/runtime/page.vue [hunk 2]
```diff
@@ -3,7 +3,7 @@
     <NuxtLayout v-if="Component" :name="layout || route.meta.layout">
       <NuxtTransition :options="route.meta.transition ?? { name: 'page', mode: 'out-in' }">
         <Suspense @pending="() => onSuspensePending(Component)" @resolve="() => onSuspenseResolved(Component)">
-          <component :is="Component" :key="route.path" />
+          <component :is="Component" />
```
Comment (pi0) [inline packages/nuxt3/src/pages/runtime/page.vue:6]: Does it means (main) page components are not cached anymore in case of client-side revisit?

Comment (toooni) [inline packages/nuxt3/src/pages/runtime/page.vue:6]: @danielroe I think there should be a value like: `:key="$route.meta.key"` here too. In our case we are changing the locale and this should result in a rerendering. An optional key would give us this possibility.

Comment (danielroe) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: would you raise a feature-request issue with an example of your use-case? :pray

Comment (toooni) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: I'll do that 👍

Comment (misaon) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: @danielroe when you removed `:key="route.path"` from page component, my page file
`./pages/[code].vue` is loaded only once, when i go from url eg. `/test-1` to `/test-2`, logic in file `./pages/[code].vue` not trigged, but router change url in adress bar. So page not redraw with fresh content from data fetch mothods. Its ok?

Comment (misaon) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: https://github.com/nuxt/framework/issues/2903

Comment (toooni) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: Yeah, it seems like this issue is bigger than my case. Having an optional `$route.meta.key` could be used to solve this issue. But with the routing strategy nuxt3 uses, there must be another/better solution which works out of the box.

Comment (misaon) [reply packages/nuxt3/src/pages/runtime/page.vue:6]: You can explicitly define key on `<NuxtPage :key="route.path" />` but page transitions wont work.

**Pull request merged**